macros/src/blanket_ref_impl.rs

```diff
@@ -0,0 +1,108 @@
+use proc_macro2::TokenStream;
+use quote::{quote, ToTokens};
+use syn::spanned::Spanned;
+use syn::{parse_quote, parse_quote_spanned};
+use syn::{Block, Ident, ImplItem, ImplItemFn, ItemTrait, Pat, Signature, TraitItem, Visibility};
+
+pub fn emit_trait_impl(trait_def: &ItemTrait, mutable: bool, stream: &mut TokenStream) {
+    let name = &trait_def.ident;
+    let impl_items = make_impl_body(&trait_def.items, mutable);
+    let implementor = if mutable { quote!(&mut T) } else { quote!(&T) };
+    quote! {
+        impl<T: #name> #name for #implementor {
+            #(#impl_items)*
+        }
+    }
+    .to_tokens(stream);
+}
+
+/// Input:
+/// ```
+/// type Foo;
+/// fn bar(&self, baz: Qux) -> Quux;
+/// ```
+///
+/// Output:
+/// ```
+/// type Foo = T::Foo;
+/// fn bar(&self, baz: Qux) -> Quux { T::bar(self, baz) }
+/// ```
+fn make_impl_body(trait_items: &[TraitItem], mutable: bool) -> Vec<ImplItem> {
+    let mut impl_items = vec![];
+
+    for item in trait_items {
+        let trait_fn = match item {
+            TraitItem::Fn(f) => f,
+            TraitItem::Type(associated_type) => {
+                let name = &associated_type.ident;
+                impl_items.push(parse_quote! {
+                    type #name = T::#name;
+                });
+                continue;
+            }
+            _ => continue,
+        };
+
+        let block = make_fn_body(&trait_fn.sig, mutable);
+
+        // Copy the signature from the trait's definition.
+        // Discard attributes and the default body.
+        let impl_fn = ImplItemFn {
+            attrs: vec![],
+            vis: Visibility::Inherited,
+            defaultness: None,
+            sig: trait_fn.sig.clone(),
+            block,
+        };
+        impl_items.push(impl_fn.into());
+    }
+
+    impl_items
+}
+
+/// Input:
+/// ```
+/// fn bar(&self, baz: Qux) -> Quux
+/// ```
+///
+/// Output:
+/// ```
+/// { T::bar(self, baz) }
```

Comment:
the input/output comments are helpful, maybe add them elsewhere too?
Maybe a high level one for blanket_ref, etc.
Perhaps use "realistic" names that sound like real AST nodes, so it's easier to categorize

AFAICT, T is the concrete node here so we implement all traits for `&IfStatement` and `&mut IfStatement`.
The mutable variant is questionable because we never need that outside the populator.

Reply:
True, the two traits we actually need the `#[blanket_mut]` *attribute* for are quite simple.
However, to implement that, we still need e.g., `impl<T: Node> Node for &mut T`, which isn't so trivial.
As long as we need to keep the mutable flavor around, we might as well have the attribute for it.

Reply:
Doc comment added to top level.